components/Markets/TwapDisplay.tsx

```diff
@@ -0,0 +1,149 @@
+  countdownEndConditions,
+  totalImpact,
+}) => {
+  const queryClient = useQueryClient();
+  const provider = useProvider();
+  const [clusterTimestamp, setClusterTimestamp] = useState<number>(0);
+  const [observedTimestamp, setObservedTimestamp] = useState<number>(0);
+  const { generateExplorerLink } = useExplorerConfiguration();
+
+  const { data: slotData } = useQuery({
+    queryKey: ['getSlot'],
+    queryFn: () => provider.connection.getSlot(),
+    staleTime: 30_000,
+  });
+  const slot = slotData ?? 0;
+
+  const timeSinceObservation = () => {
+    const diff = clusterTimestamp - observedTimestamp;
+    if (diff > 864_000) {
+      return 'A long time ago';
+    }
+    if (diff > 86_400) {
+      const _diff = diff / 86_400;
+      return `${_diff.toFixed(0)}+ days ago`;
+    }
+    if (diff > 3_600) {
+      // hours
+      const _diff = diff / 3_600;
+      return `${_diff.toFixed(0)}+ hours ago`;
+    }
+    if (diff > 60) {
+      // minutes
+      const _diff = diff / 60;
+      return `${_diff.toFixed(0)}+ minutes ago`;
+    }
+    return `${diff} seconds ago`;
+  };
+
+  const getClusterTimestamp = async () => {
+    let _clusterTimestamp: number = 0;
+    if (slot !== 0) {
+      _clusterTimestamp = await queryClient.fetchQuery({
+        queryKey: [`getBlockTime-${slot}`],
+        queryFn: () => provider.connection.getBlockTime(slot),
+        staleTime: 30_000,
+      });
+    }
+    const _observedTimestamp = await queryClient.fetchQuery({
+      queryKey: [`getBlockTime-${lastObservedSlot}`],
+      queryFn: () => provider.connection.getBlockTime(lastObservedSlot),
+      staleTime: 30_000,
+    });
+    if (_clusterTimestamp) {
+      setClusterTimestamp(_clusterTimestamp);
+    }
+    if (_observedTimestamp) {
+      setObservedTimestamp(_observedTimestamp);
+    }
+  };
+
+  useEffect(() => {
+    if ((!clusterTimestamp || clusterTimestamp === 0) && slot) {
+      getClusterTimestamp();
+    }
+  }, [slot]);
+
+  return (
+    <Group justify="center" align="center">
+      <Stack gap={0} pb="1rem" align="center">
+        <Group gap={3} justify="center" align="center">
+          <Text fw="bold" size="md">
+            ${numeral(twap).format(NUMERAL_FORMAT)}
+          </Text>
+          <Text size="sm">TWAP</Text>
+        </Group>
+        <Text size="xs">${numeral(midPrice).format(NUMERAL_FORMAT)} (mid)</Text>
+      </Stack>
+      <HoverCard position="top">
+        <HoverCard.Target>
+          <IconInfoCircle strokeWidth={1.3} />
+        </HoverCard.Target>
+        <HoverCard.Dropdown w="22rem">
+          <Stack>
+            <Text>{countdownEndConditions}</Text>
+            <Text>
+              Last observed price (for TWAP calculation) $
+              {numeral(lastObservationValue).format(NUMERAL_FORMAT)}
+            </Text>
+            <Text size="xs">
+              Last observed at
+              <br />
+              slot {lastObservedSlot} | {slot - lastObservedSlot} slots behind cluster
```

Comment:
So here we've got a stale slot. This is a bit tricky, the solution I can think of is adding a slotSubscribe and just keeping cluster time and slot height in a state system such that it's referenced from websocket and we always have tip and time to reference in here, otherwise we end up with a fetch and stale data...

<img width="282" alt="Screenshot 2024-03-18 at 10 40 00 PM" src="https://github.com/metaDAOproject/meta-dao-frontend/assets/1007667/e9b9a981-1fa0-4722-8834-d1a7aba91bd1">

Up to you. We can remove everything but the last updated slot, however I think it is useful to know the cluster state....

Reply:
This is stale now because the rest is updating I guess? Does that make it technically inaccurate or just stale but accurate?

If it's inaccurate I guess we should definitely do it. If it's still accurate I am not sure. I would see what you think regarding adding this functionality to the old UI and if you think it's worth the extra time in the case it is accurate but just not updating.

 When you say remove everything but the last updated slot... you mean remove the "Last observed at..." section?

Reply:
@R-K-H just added the change for reactive cluster slot and it should be looking  good now.